### ISSUE TITLE
UrlHelper in Blazor component works on first load, but not on second.

### ISSUE THREAD

**Issue by JHBonarius** [dotnet/aspnetcore#31097]
### Describe the bug
Let me start with something that might be a different bug: Every time I load a Blazor Component, it seems to be loaded(/rendered) twice. I.e. if I set a breakpoint in `OnInitialized`, it is triggered twice. Is this normal? edit: ah found it: #13448

Anyhow, I'm trying to use MVC routing in a Blazor Component, in order to prevent hard coded routes. I found the taghelpers are not supported in Blazor, so I tried something else.
After looking at [this](https://benfoster.io/blog/injecting-urlhelper-in-aspnet-core-mvc/) I added the Following to my `Startup.cs`
```c#
services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();
services.AddScoped(x =>
{
    var actionContextAccessor = x.GetRequiredService<IActionContextAccessor>();
    var factory = x.GetRequiredService<IUrlHelperFactory>();
    return factory.GetUrlHelper(actionContextAccessor.ActionContext);
});
```

Then in my Blazor component I can use:
```xml
@using Microsoft.AspNetCore.Mvc
@inject IUrlHelper UrlHelper
...
<a href="@UrlHelper.ActionLink(controller: "Home", action: "Index")">Link</a>
...
```

And this actually produces the correct html on first load. (when `actionContextAccessor.ActionContext` is `Microsoft.AspNetCore.Mvc.RazorPages.PageContext`)
```html
<a href="https://localhost:44364/Home/Index" b-ai7am37k0o="">Link</a>
```

...however, somehow a _second_ load with a new scope is initiated, so the scoped UrlHelper instance it recreated, but now `actionContextAccessor.ActionContext` is `null`... that causes an exception in `factory.GetUrlHelper`.

edit 2: I found the double load is caused by prerendering. But if I disable that (by setting it to `RenderMode.Server`), `actionContextAccessor.ActionContext` is always `null`, which makes the problem worse.

### To Reproduce
Take the default Blazor server-side project and add MVC. I.e. `services.AddControllersWithViews();`, `endpoints.MapControllers();` and add a controller
```
    public class HomeController : Controller
    {
        [Route("Home/Index")]
        public IActionResult Index() { return View(); }
    }
```
And some (empty) view.

Then add the code in the bug to a Blazor Component (I used NavMenu.razor)

edit: to be sure, I just put it all in [a github repository](https://github.com/JHBonarius/BlazorServerDoubleLoad) for you to quickly get.

### Exceptions (if any)
```
System.ArgumentNullException: Value cannot be null. (Parameter 'context')
         at Microsoft.AspNetCore.Mvc.Routing.UrlHelperFactory.GetUrlHelper(ActionContext context)
         at BlazorServerDefault.Startup.<>c.<ConfigureServices>b__4_1(IServiceProvider x) in D:\Cs\Scrapbook\BlazorServerDefault\Startup.cs:line 45
```

### Further technical details
- ASP.NET Core version 5.0.0
- Include the output of `dotnet --info`
```
.NET SDK (reflecting any global.json):
 Version:   5.0.201
 Commit:    a09bd5c86c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.201\

Host (useful for support):
  Version: 5.0.4
  Commit:  f27d337295

.NET SDKs installed:
  3.1.407 [C:\Program Files\dotnet\sdk]
  5.0.101 [C:\Program Files\dotnet\sdk]
  5.0.200-preview.21077.7 [C:\Program Files\dotnet\sdk]
  5.0.200 [C:\Program Files\dotnet\sdk]
  5.0.201 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version VS 2019 16.9.2 and 16.10.0 preview 1.0

**Comment by javiercn** [comment]
@JHBonarius thanks for contacting us.

It works because on first load it's prerendering. URLHelper relies on HttpContext, which is discouraged inside Blazor applications. In general you should avoid any ASP.NET Core component that directly or indirectly depends on HttpContext or a service that is only available within the context of an HTTP request.

Our recommendation is that instead you pass in the data as parameters to the root component.

**Comment by JHBonarius** [comment]
@javiercn I don't understand what you mean. And I don't think this is resolved.

I want to be able to use the UrlHelper or any dynamic method that resolves the path to a MVC controller method. *These are not fixed in our design*. We **don't** want hard-coded paths for mainenance
I.e. now it's [Route("Home/Index")], but in future it could become [Route("Nananana/Batman")] or so. I mean that's the whole reason there is something like the taghelpers.
And you must understand that this is a simplified design. In our real system we have 30+ controller with each many methods.

I tried to implement what I think you mean. To `_Host.cshtml` I added a model
```c#
    public class _HostModel : PageModel
    {
        public IUrlHelper UrlHelper { get; init; }

        public _HostModel(IUrlHelper urlHelper)
        {
            UrlHelper = urlHelper;
        }
    }
```
Then I modified `_Host.cshtml`.
```html
    <component type="typeof(App)" param-UrlHelper="Model.UrlHelper" render-mode="Server" />
```
To `App.razor` I added:
```c#
@code{
    [Parameter]
    public IUrlHelper UrlHelper { get; set; }
}
```
But here I'm stuck. How do I conditionally pass this parameter to the Component found by the `Router`?

Furthermore, tbh I think adding a Parameter `UrlHelper` to each and every components and passing it along the hierarchy seems insanity. That's what we have DI for nowadays, right?
